1112-i386-txt-Add-Intel-TXT-ACM-module-support.patch

```diff
@@ -0,0 +1,596 @@
++}
++
++static struct grub_txt_acm_info_table*
++get_acmod_info_table (struct grub_txt_acm_header* hdr)
++{
++  grub_uint32_t user_area_off;
++
++  /* Overflow? */
++  if ( plus_overflow_u32 (hdr->header_len, hdr->scratch_size) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM header length plus scratch size overflows"));
++      return NULL;
++    }
++
++  if ( multiply_overflow_u32 ((hdr->header_len + hdr->scratch_size), 4) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM header length and scratch size in bytes overflows"));
++      return NULL;
++    }
++
++  /*
++   * This fn assumes that the ACM has already passed at least the initial
++   * is_acmod() checks.
++   */
++
++  user_area_off = (hdr->header_len + hdr->scratch_size) * 4;
++
++  /* Overflow? */
++  if ( plus_overflow_u32 (user_area_off, sizeof(struct grub_txt_acm_info_table)) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("user_area_off plus acm_info_table_t size overflows"));
++      return NULL;
++    }
++
++  /* Check that table is within module. */
++  if ( user_area_off + sizeof(struct grub_txt_acm_info_table) > hdr->size * 4 )
++    {
++      /* TODO: Is (grub_uint32_t) correct??? */
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM info table size too large: %x > %x"),
++		  user_area_off + (grub_uint32_t)sizeof(struct grub_txt_acm_info_table), hdr->size * 4);
++      return NULL;
++    }
++
++  /* Overflow? */
++  if ( plus_overflow_u32 ((grub_uint32_t)(unsigned long)hdr, user_area_off) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("hdr plus user_area_off overflows"));
++      return NULL;
++    }
++
++    return (struct grub_txt_acm_info_table *)((unsigned long)hdr + user_area_off);
++}
++
++static struct grub_txt_acm_chipset_id_list*
++get_acmod_chipset_list (struct grub_txt_acm_header *hdr)
++{
++  struct grub_txt_acm_info_table *info_table;
++  grub_uint32_t size, id_list_off;
++  struct grub_txt_acm_chipset_id_list *chipset_id_list;
++
++  /* This fn assumes that the ACM has already passed the is_acmod() checks */
++
++  info_table = get_acmod_info_table (hdr);
++  if ( !info_table )
++    return NULL;
++  id_list_off = info_table->chipset_id_list;
++
++  size = hdr->size * 4;
++
++  /* Overflow? */
++  if ( plus_overflow_u32 (id_list_off, sizeof(struct grub_txt_acm_chipset_id)) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("id_list_off plus acm_chipset_id_t size overflows"));
++      return NULL;
++    }
++
++  /* Check that chipset id table is w/in ACM */
++  if ( id_list_off + sizeof(struct grub_txt_acm_chipset_id) > size )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("chipset id list is too big: %x"), id_list_off);
++      return NULL;
++    }
++
++  /* Overflow? */
++  if ( plus_overflow_u32 ((grub_uint32_t)(unsigned long)hdr, id_list_off) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("hdr plus id_list_off overflows"));
++      return NULL;
++    }
```

Comment:
As before, checking for overflow of `pointer + offset`, when you've already checked that the offset is inbounds for the object the pointer points to, seems like a code smell.